### ISSUE TITLE
fparser2 recursion error

### ISSUE THREAD

**Issue by rupertford** [stfc/fparser#167]
Some UM routines cause fparser2 to fail with:

`RuntimeError: maximum recursion depth exceeded`

As noted by @MartinDix, you can increase the default recursion depth and fparser2 will work. However I thought I'd check whether there is anything going on (probably due to lax matching) that could be improved.

One cut down example that fails is the following ...

```
SUBROUTINE test()

  infox(       1:     447) =  (/                                    &
   4.814e-12, 4.947e-12, 5.080e-12, 5.213e-12, 5.346e-12, 5.479e-12,&
   5.612e-12, 5.745e-12, 5.878e-12, 6.011e-12, 6.139e-12, 6.272e-12,&
   6.434e-12, 6.628e-12, 6.823e-12, 7.013e-12, 7.207e-12, 7.397e-12,&
   7.592e-12, 7.782e-12, 7.972e-12, 8.162e-12, 8.352e-12, 8.542e-12,&
   8.770e-12, 9.035e-12, 9.297e-12, 9.558e-12, 9.819e-12, 1.008e-11,&
   1.034e-11, 1.060e-11, 1.086e-11, 1.112e-11, 1.138e-11, 1.164e-11,&
   1.193e-11, 1.226e-11, 1.258e-11, 1.291e-11, 1.323e-11, 1.355e-11,&
   1.387e-11, 1.420e-11, 1.451e-11, 1.484e-11, 1.516e-11, 1.548e-11,&
   1.583e-11, 1.624e-11, 1.664e-11, 1.704e-11, 1.743e-11, 1.783e-11,&
   1.823e-11, 1.863e-11, 1.903e-11, 1.942e-11, 1.981e-11, 2.021e-11,&
   2.065e-11, 2.115e-11, 2.165e-11, 2.214e-11, 2.264e-11, 2.314e-11,&
   2.363e-11, 2.412e-11, 2.461e-11, 2.511e-11, 2.560e-11, 2.609e-11,&
   2.661e-11, 2.716e-11, 2.771e-11, 2.826e-11, 2.881e-11, 2.936e-11,&
   2.990e-11, 3.045e-11, 3.099e-11, 3.154e-11, 3.208e-11, 3.262e-11,&
   3.315e-11, 3.364e-11, 3.415e-11, 3.465e-11, 3.515e-11, 3.565e-11,&
   3.615e-11, 3.664e-11, 3.714e-11, 3.763e-11, 3.813e-11, 3.862e-11,&
   3.912e-11, 3.963e-11, 4.013e-11, 4.063e-11, 4.113e-11, 4.164e-11,&
   4.213e-11, 4.264e-11, 4.314e-11, 4.363e-11, 4.413e-11, 4.462e-11,&
   4.521e-11, 4.588e-11, 4.655e-11, 4.723e-11, 4.790e-11, 4.857e-11,&
   4.924e-11, 4.990e-11, 5.057e-11, 5.124e-11, 5.190e-11, 5.256e-11,&
   5.333e-11, 5.420e-11, 5.508e-11, 5.595e-11, 5.682e-11, 5.769e-11,&
   5.856e-11, 5.943e-11, 6.029e-11, 6.115e-11, 6.201e-11, 6.287e-11,&
   6.386e-11, 6.496e-11, 6.606e-11, 6.716e-11, 6.826e-11, 6.935e-11,&
   7.045e-11, 7.154e-11, 7.263e-11, 7.372e-11, 7.480e-11, 7.589e-11,&
   7.711e-11, 7.846e-11, 7.981e-11, 8.116e-11, 8.251e-11, 8.385e-11,&
   8.519e-11, 8.653e-11, 8.787e-11, 8.920e-11, 9.053e-11, 9.186e-11,&
   9.332e-11, 9.493e-11, 9.653e-11, 9.812e-11, 9.972e-11, 1.013e-10,&
   1.029e-10, 1.045e-10, 1.061e-10, 1.076e-10, 1.092e-10, 1.108e-10,&
   1.125e-10, 1.143e-10, 1.161e-10, 1.179e-10, 1.197e-10, 1.215e-10,&
   1.233e-10, 1.251e-10, 1.269e-10, 1.287e-10, 1.305e-10, 1.323e-10,&
   1.342e-10, 1.362e-10, 1.382e-10, 1.402e-10, 1.422e-10, 1.442e-10,&
   1.462e-10, 1.482e-10, 1.502e-10, 1.522e-10, 1.542e-10, 1.562e-10,&
   1.583e-10, 1.606e-10, 1.629e-10, 1.651e-10, 1.674e-10, 1.697e-10,&
   1.719e-10, 1.742e-10, 1.764e-10, 1.787e-10, 1.809e-10, 1.832e-10,&
   1.856e-10, 1.882e-10, 1.908e-10, 1.933e-10, 1.959e-10, 1.985e-10,&
   2.010e-10, 2.036e-10, 2.062e-10, 2.087e-10, 2.112e-10, 2.138e-10,&
   2.166e-10, 2.196e-10, 2.226e-10, 2.255e-10, 2.285e-10, 2.315e-10,&
   2.345e-10, 2.374e-10, 2.404e-10, 2.434e-10, 2.463e-10, 2.493e-10,&
   2.524e-10, 2.558e-10, 2.592e-10, 2.626e-10, 2.660e-10, 2.694e-10,&
   2.727e-10, 2.761e-10, 2.795e-10, 2.828e-10, 2.862e-10, 2.895e-10,&
   2.930e-10, 2.967e-10, 3.004e-10, 3.041e-10, 3.078e-10, 3.115e-10,&
   3.152e-10, 3.189e-10, 3.225e-10, 3.262e-10, 3.298e-10, 3.335e-10,&
   3.374e-10, 3.415e-10, 3.457e-10, 3.499e-10, 3.540e-10, 3.582e-10,&
   3.623e-10, 3.664e-10, 3.705e-10, 3.746e-10, 3.787e-10, 3.828e-10,&
   3.873e-10, 3.921e-10, 3.968e-10, 4.016e-10, 4.063e-10, 4.111e-10,&
   4.158e-10, 4.205e-10, 4.252e-10, 4.299e-10, 4.346e-10, 4.393e-10,&
   4.443e-10, 4.495e-10, 4.547e-10, 4.599e-10, 4.651e-10, 4.703e-10,&
   4.755e-10, 4.807e-10, 4.858e-10, 4.910e-10, 4.961e-10, 5.013e-10,&
   5.063e-10, 5.112e-10, 5.162e-10, 5.211e-10, 5.260e-10, 5.309e-10,&
   5.358e-10, 5.406e-10, 5.455e-10, 5.504e-10, 5.552e-10, 5.601e-10,&
   5.650e-10, 5.699e-10, 5.749e-10, 5.798e-10, 5.848e-10, 5.897e-10,&
   5.946e-10, 5.995e-10, 6.044e-10, 6.093e-10, 6.141e-10, 6.190e-10,&
   6.238e-10, 6.284e-10, 6.330e-10, 6.376e-10, 6.422e-10, 6.468e-10,&
   6.514e-10, 6.560e-10, 6.606e-10, 6.651e-10, 6.697e-10, 6.742e-10,&
   6.790e-10, 6.840e-10, 6.890e-10, 6.940e-10, 6.990e-10, 7.040e-10,&
   7.090e-10, 7.140e-10, 7.189e-10, 7.239e-10, 7.288e-10, 7.337e-10,&
   7.379e-10, 7.414e-10, 7.448e-10, 7.482e-10, 7.516e-10, 7.550e-10,&
   7.584e-10, 7.618e-10, 7.652e-10, 7.686e-10, 7.720e-10, 7.753e-10,&
   7.790e-10, 7.829e-10, 7.868e-10, 7.906e-10, 7.945e-10, 7.984e-10,&
   8.022e-10, 8.061e-10, 8.099e-10, 8.138e-10, 8.176e-10, 8.214e-10,&
   8.251e-10, 8.286e-10, 8.321e-10, 8.357e-10, 8.392e-10, 8.427e-10,&
   8.462e-10, 8.497e-10, 8.532e-10, 8.567e-10, 8.602e-10, 8.636e-10,&
   8.668e-10, 8.696e-10, 8.724e-10, 8.752e-10, 8.780e-10, 8.808e-10,&
   8.836e-10, 8.863e-10, 8.891e-10, 8.919e-10, 8.947e-10, 8.974e-10,&
   9.006e-10, 9.041e-10, 9.077e-10, 9.112e-10, 9.148e-10, 9.183e-10,&
   9.219e-10, 9.254e-10, 9.289e-10, 9.324e-10, 9.359e-10, 9.394e-10,&
   9.430e-10, 9.467e-10, 9.504e-10, 9.541e-10, 9.578e-10, 9.614e-10,&
   9.651e-10, 9.687e-10, 9.724e-10, 9.760e-10, 9.796e-10, 9.833e-10,&
   9.869e-10, 9.906e-10, 9.943e-10, 9.980e-10, 1.002e-09, 1.005e-09,&
   1.009e-09, 1.013e-09, 1.016e-09, 1.020e-09, 1.024e-09, 1.027e-09,&
   1.031e-09, 1.036e-09, 1.041e-09, 1.046e-09, 1.051e-09, 1.056e-09,&
   1.061e-09, 1.065e-09, 1.070e-09, 1.075e-09, 1.080e-09, 1.085e-09,&
   1.089e-09, 1.094e-09, 1.099e-09, 1.104e-09, 1.108e-09, 1.113e-09,&
   1.118e-09, 1.123e-09, 1.127e-09, 1.132e-09, 1.137e-09, 1.141e-09,&
   1.145e-09, 1.148e-09, 1.152e-09/)

END SUBROUTINE
```

**Comment by rupertford** [comment]
The following also fails ...

```
MODULE test

  REAL (RealK), PARAMETER :: xxxxx(364) = (/                            &
      0.175639712E-07,0.178048731E-06,0.168366457E-05,0.830875319E-04,  &
      0.424841216E-03,0.134550228E-02,0.564003231E-02,0.352066791E-01,  &
      0.310019938E+00,0.335244996E+01,0.620707596E+02,0.153459831E+04,  &
      0.364951217E+05,0.281792180E+06,                                  &
      0.289265901E-07,0.291576019E-06,0.271041663E-05,0.132438851E-03,  &
      0.688127390E-03,0.213266164E-02,0.865449254E-02,0.499274729E-01,  &
      0.416031184E+00,0.410411090E+01,0.629682670E+02,0.153704730E+04,  &
      0.364647305E+05,0.281552401E+06,                                  &
      0.476186919E-07,0.477868738E-06,0.437822451E-05,0.210134182E-03,  &
      0.111720422E-02,0.339495593E-02,0.134372812E-01,0.729199225E-01,  &
      0.589538979E+00,0.537530220E+01,0.652374815E+02,0.154106681E+04,  &
      0.364147695E+05,0.281158185E+06,                                  &
      0.785071475E-07,0.784098283E-06,0.707359175E-05,0.334181799E-03,  &
      0.181569792E-02,0.542792491E-02,0.207562168E-01,0.105427043E+00,  &
      0.814811496E+00,0.688752698E+01,0.718173593E+02,0.154767126E+04,  &
      0.363324674E+05,0.280508852E+06,                                  &
      0.129212512E-06,0.128649619E-05,0.115041745E-04,0.532659395E-03,  &
      0.295704217E-02,0.871932896E-02,0.321417049E-01,0.161457594E+00,  &
      0.114843349E+01,0.908448115E+01,0.839321624E+02,0.158637475E+04,  &
      0.361976310E+05,0.279445144E+06,                                  &
      0.212881630E-06,0.210841622E-05,0.187945911E-04,0.856163994E-03,  &
      0.482756508E-02,0.140310026E-01,0.502060298E-01,0.242381397E+00,  &
      0.159347322E+01,0.120912670E+02,0.999223285E+02,0.157730138E+04,  &
      0.359768094E+05,0.277703439E+06,                                  &
      0.350738684E-06,0.345147582E-05,0.305922018E-04,0.137147692E-02,  &
      0.787684770E-02,0.226697077E-01,0.789732432E-01,0.371881781E+00,  &
      0.240823743E+01,0.173153155E+02,0.121006994E+03,0.180452778E+04,  &
      0.356175474E+05,0.274870907E+06,                                  &
      0.578087452E-06,0.565959693E-05,0.496798293E-04,0.220247298E-02,  &
      0.128544768E-01,0.366165849E-01,0.124979071E+00,0.580745403E+00,  &
      0.362567398E+01,0.243894617E+02,0.143455145E+03,0.249186140E+04,  &
      0.350368844E+05,0.270295087E+06,                                  &
      0.952560665E-06,0.928228519E-05,0.810582735E-04,0.356148645E-02,  &
      0.209581372E-01,0.591818354E-01,0.199644176E+00,0.900487875E+00,  &
      0.544952114E+01,0.340343369E+02,0.208411370E+03,0.278939487E+04,  &
      0.341126320E+05,0.263017980E+06,                                  &
      0.157058733E-05,0.152202291E-04,0.132309885E-03,0.578249214E-02,  &
      0.342784044E-01,0.958326558E-01,0.320609549E+00,0.140700166E+01,  &
      0.828197496E+01,0.503833905E+02,0.297514385E+03,0.386838360E+04,  &
      0.326695251E+05,0.251671080E+06,                                  &
      0.258797948E-05,0.249859659E-04,0.215456608E-03,0.935563117E-02,  &
      0.561512099E-01,0.155548595E+00,0.517029940E+00,0.220861092E+01,  &
      0.125395515E+02,0.712262934E+02,0.443228175E+03,0.452409719E+04,  &
      0.341069595E+05,0.238340983E+06,                                  &
      0.426488182E-05,0.410570282E-04,0.350034250E-03,0.151799120E-01,  &
      0.916676371E-01,0.252995611E+00,0.828255578E+00,0.355063251E+01,  &
      0.192792370E+02,0.101384545E+03,0.614100075E+03,0.646494493E+04,  &
      0.399466981E+05,0.244811585E+06,                                  &
      0.703002623E-05,0.674204555E-04,0.569873387E-03,0.246181197E-01,  &
      0.149973906E+00,0.412587798E+00,0.133210731E+01,0.564163334E+01,  &
      0.294800800E+02,0.152079426E+03,0.926709704E+03,0.942944249E+04,  &
      0.410885683E+05,0.217664796E+06,                                  &
      0.115854159E-04,0.110669039E-03,0.926062499E-03,0.400616110E-01,  &
      0.244828814E+00,0.666431333E+00,0.213558017E+01,0.884671389E+01,  &
      0.441602827E+02,0.230110182E+03,0.139839658E+04,0.134659292E+05,  &
      0.462675313E+05,0.201867848E+06,                                  &
      0.190907674E-04,0.181309923E-03,0.150296677E-02,0.649591464E-01,  &
      0.399606739E+00,0.107729034E+01,0.343038530E+01,0.138583877E+02,  &
      0.658921633E+02,0.341064697E+03,0.199935151E+04,0.184864702E+05,  &
      0.580124909E+05,0.171374217E+06,                                  &
      0.314598350E-04,0.297341589E-03,0.241387462E-02,0.104034787E+00,  &
      0.651981252E+00,0.174137843E+01,0.547296471E+01,0.213017587E+02,  &
      0.100005817E+03,0.506599314E+03,0.288314249E+04,0.208328613E+05,  &
      0.543287387E+05,0.130993190E+06,                                  &
      0.518488658E-04,0.485979864E-03,0.381655435E-02,0.165055611E+00,  &
      0.105785507E+01,0.278930779E+01,0.842545670E+01,0.315594942E+02,  &
      0.146905037E+03,0.734536293E+03,0.422540329E+04,0.224370729E+05,  &
      0.424878217E+05,0.907544273E+05,                                  &
      0.854597325E-04,0.789425904E-03,0.599189247E-02,0.257620183E+00,  &
      0.170532552E+01,0.444188072E+01,0.128443866E+02,0.476233760E+02,  &
      0.216221872E+03,0.108002716E+04,0.573419583E+04,0.224711078E+05,  &
      0.327907876E+05,0.608935859E+05,                                  &
      0.140854171E-03,0.126509538E-02,0.932837192E-02,0.399201229E+00,  &
      0.274126326E+01,0.693659937E+01,0.191330372E+02,0.703062416E+02,  &
      0.309289041E+03,0.154966727E+04,0.668899422E+04,0.191328726E+05,  &
      0.261309718E+05,0.431114680E+05,                                  &
      0.232114521E-03,0.201988410E-02,0.145574602E-01,0.604777002E+00,  &
      0.435903411E+01,0.107851230E+02,0.287619709E+02,0.995166936E+02,  &
      0.443498834E+03,0.203621109E+04,0.704403565E+04,0.164119575E+05,  &
      0.215081485E+05,0.294716721E+05,                                  &
      0.380316936E-03,0.320387682E-02,0.225903086E-01,0.907773838E+00,  &
      0.688470723E+01,0.165564632E+02,0.431346391E+02,0.144172448E+03,  &
      0.635047711E+03,0.237632781E+04,0.626716979E+04,0.148186209E+05,  &
      0.183561801E+05,0.211105775E+05,                                  &
      0.617711325E-03,0.512415009E-02,0.340677704E-01,0.134011810E+01,  &
      0.109396878E+02,0.244685829E+02,0.630721092E+02,0.202080534E+03,  &
      0.801894508E+03,0.233826111E+04,0.498818597E+04,0.136676022E+05,  &
      0.164042748E+05,0.175707567E+05,                                  &
      0.101120703E-02,0.799308562E-02,0.502085993E-01,0.197951492E+01,  &
      0.169992752E+02,0.358621095E+02,0.933044648E+02,0.280488296E+03,  &
      0.912118152E+03,0.204514079E+04,0.414887086E+04,0.128644668E+05,  &
      0.146341003E+05,0.153315364E+05,                                  &
      0.163543015E-02,0.123173064E-01,0.713920952E-01,0.279958561E+01,  &
      0.263591034E+02,0.528139374E+02,0.133439649E+03,0.354014551E+03,  &
      0.892401336E+03,0.164237224E+04,0.431880510E+04,0.112413811E+05,  &
      0.124497167E+05,0.128833816E+05,                                  &
      0.260892696E-02,0.185526783E-01,0.943293030E-01,0.389381831E+01,  &
      0.395668174E+02,0.788572857E+02,0.175239663E+03,0.399393824E+03,  &
      0.771597646E+03,0.127553915E+04,0.444506619E+04,0.949084154E+04,  &
      0.101602596E+05,0.103854380E+05,                                  &
      0.411523928E-02,0.275285205E-01,0.108651131E+00,0.496714742E+01,  &
      0.563999636E+02,0.111069005E+03,0.214571085E+03,0.394996342E+03,  &
      0.609849719E+03,0.113806327E+04,0.443190119E+04,0.751737709E+04,  &
      0.784369028E+04,0.794299694E+04 /)

END MODULE test
```

**Comment by rupertford** [comment]
The other two UM files with the recursion problem fail with the same long array initialisation.

**Comment by pelson** [comment]
I just checked and fparser two handles large arrays of INTEGERS (I tested to an array of size 4000, and could definitely see it slowing down, but no :boom:). It is the ``REAL`` that is catching it out...

Update: fixed by #182.
